tomb-common/src/types/blockstore/car/carv1/mod.rs

```diff
@@ -0,0 +1,463 @@
+// Modules
+pub mod carv1blockstore;
+pub(crate) mod v1block;
+pub(crate) mod v1header;
+pub(crate) mod v1index;
+
+// Code
+use anyhow::Result;
+use serde::{Deserialize, Serialize};
+use std::{
+    cell::RefCell,
+    collections::HashMap,
+    io::{Read, Seek, SeekFrom, Write},
+};
+use wnfs::libipld::Cid;
+
+use self::{v1block::V1Block, v1header::V1Header, v1index::V1Index};
+
+use super::carv2::V2_PH_SIZE;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub(crate) struct CarV1 {
+    pub header: V1Header,
+    pub index: RefCell<V1Index>,
+    pub(crate) read_header_len: RefCell<u64>,
+}
+
+impl CarV1 {
+    /// Read in a CARv1 object, assuming the Reader is already seeked to the first byte of the CARv1
+    pub(crate) fn read_bytes<R: Read + Seek>(mut r: R) -> Result<Self> {
+        // Track the part of the stream where the V1Header starts
+        let header_start = r.stream_position()?;
+        // Read the V1Header
+        let header = V1Header::read_bytes(&mut r)?;
+        // Determine the length of the header that we just read
+        let read_header_len = RefCell::new(r.stream_position()? - header_start);
+        // Generate an index
+        let index = RefCell::new(V1Index::read_bytes(&mut r)?);
+        Ok(Self {
+            header,
+            index,
+            read_header_len,
+        })
+    }
+
+    /// Write out a CARv1 object, assuming the Writer is already seeked to the first byte of the CARv1
+    pub(crate) fn write_bytes<R: Read + Seek, W: Write + Seek>(
+        &self,
+        mut r: R,
+        mut w: W,
+    ) -> Result<()> {
+        // Save our starting point
+        let carv1_start = r.stream_position()?;
+        w.seek(SeekFrom::Start(carv1_start))?;
```

Comment:
i believe that this is something we can fix by traversing backwards through the file. if we already have an map of blocks and their offsets built, and start with the blocks written furthest into the file, then there is no chance of overwriting existing data. either the header has increased in size or remained fixed in size. in the former case, all blocks are being shifted further into the file, so reading a 100 byte block and skipping 50 bytes of offset into it to overwrite it wont matter if there are no blocks succeeding it. in the latter case, no writing even needs to be performed. 
that might still sound unsafe to you, but theres no longer a chance of destroying block data, even if we corrupt the file format. 
we could likely implement this with the existing v1 index, but id rather wait until we have a v2 index parser, as it will make the v1 index essentially obsolete for our use case.